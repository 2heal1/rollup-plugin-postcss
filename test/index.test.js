import path from 'path'
import fs from 'fs-extra'
import { rollup } from 'rollup'
import postcss from '../src'

function fixture(...args) {
  return path.join(__dirname, 'fixtures', ...args)
}

beforeAll(() => fs.remove(fixture('dist')))

async function write({
  input,
  dirname,
  options
}) {
  dirname = fixture('dist', dirname)
  const bundle = await rollup({
    input: fixture(input),
    plugins: [
      postcss(options)
    ]
  })
  await bundle.write({
    format: 'cjs',
    file: path.join(dirname, 'bundle.js')
  })
  const cssCodePath = path.join(dirname, 'bundle.css')
  const cssMapPath = path.join(dirname, 'bundle.css.map')
  const jsCodePath = path.join(dirname, 'bundle.js')
  return {
    jsCode() {
      return fs.readFile(jsCodePath, 'utf8')
    },
    cssCode() {
      return fs.readFile(cssCodePath, 'utf8')
    },
    cssMap() {
      return fs.readFile(cssMapPath, 'utf8')
    },
    hasCssFile() {
      return fs.pathExists(cssCodePath)
    },
    hasCssMapFile() {
      return fs.pathExists(cssMapPath)
    }
  }
}

function snapshot({
  title,
  input,
  options = {}
}) {
  test(title, async () => {
    const res = await write({
      input,
      dirname: title,
      options
    })

    expect(await res.jsCode()).toMatchSnapshot('js code')

    if (options.extract) {
      expect(await res.hasCssFile()).toBe(true)
      expect(await res.cssCode()).toMatchSnapshot('css code')
    }

    const sourceMap = options && options.sourceMap
    if (sourceMap === 'inline') {
      expect(await res.hasCssMapFile()).toBe(false)
    } else if (sourceMap === true) {
      expect(await res.hasCssMapFile()).toBe(Boolean(options.extract))
      if (options.extract) {
        expect(await res.cssMap()).toMatchSnapshot('css map')
      }
    }
  })
}

snapshot({
  title: 'simple',
  input: 'simple/index.js'
})

snapshot({
  title: 'extract',
  input: 'simple/index.js',
  options: {
    extract: true
  }
})

snapshot({
  title: 'minimize:inject',
  input: 'simple/index.js',
  options: {
    minimize: true
  }
})

snapshot({
  title: 'minimize:extract',
  input: 'simple/index.js',
  options: {
    minimize: true,
    extract: true
  }
})

snapshot({
  title: 'modules:inject',
  input: 'css-modules/index.js',
  options: {
    modules: true
  }
})

snapshot({
  title: 'modules:named-exports',
  input: 'named-exports/index.js',
  options: {
    modules: true,
    namedExports: true
  }
})

snapshot({
  title: 'modules:extract',
  input: 'css-modules/index.js',
  options: {
    modules: true,
    extract: true
  }
})

snapshot({
  title: 'sourcemap:true',
  input: 'simple/index.js',
  options: {
    sourceMap: true
  }
})

snapshot({
  title: 'extract::sourcemap:true',
  input: 'simple/index.js',
  options: {
    sourceMap: true,
    extract: true
  }
})

// inline is actually broken for now
snapshot({
  title: 'sourcemap:inline',
  input: 'simple/index.js',
  options: {
    sourceMap: 'inline'
  }
})

snapshot({
  title: 'extract::sourcemap:inline',
  input: 'simple/index.js',
  options: {
    sourceMap: 'inline',
    extract: true
  }
})

snapshot({
  title: 'inject:top',
  input: 'simple/index.js',
  options: {
    inject: {
      insertAt: 'top'
    }
  }
})

snapshot({
  title: 'sass',
  input: 'sass/index.js',
  options: {
    use: [
      'sass'
    ]
  }
})

snapshot({
  title: 'sass:sourcemap',
  input: 'sass/index.js',
  options: {
    sourceMap: true,
    use: [
      'sass'
    ]
  }
})

snapshot({
  title: 'postcss-config',
  input: 'postcss-config/index.js'
})

snapshot({
  title: 'sass:modules',
  input: 'sass-modules/index.js',
  options: {
    modules: true,
    use: [
      'sass'
    ]
  }
})

snapshot({
  title: 'skip-loader',
  input: 'simple/index.js',
  options: {
    use: ['loader'],
    loaders: [
      {
        name: 'loader',
        test: /\.random$/,
        process() {
          return 'lol'
        }
      }
    ]
  }
})

snapshot({
  title: 'postcss-options',
  input: 'postcss-options/index.js',
  options: {
    plugins: [
      require('autoprefixer')()
    ]
  }
})

snapshot({
  title: 'inject:false',
  input: 'simple/index.js',
  options: {
    inject: false
  }
})
