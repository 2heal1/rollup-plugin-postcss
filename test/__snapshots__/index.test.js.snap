// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract: css code 1`] = `
"body {
  color: red;
}

.bar {
  color: red;
}
"
`;

exports[`extract: js code 1`] = `
"'use strict';

console.log(undefined, undefined);
"
`;

exports[`extract::sourcemap:inline: css code 1`] = `
"body {
  color: red;
}

.bar {
  color: red;
}

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInRlc3QvZml4dHVyZXMvZGlzdC9leHRyYWN0Ojpzb3VyY2VtYXA6aW5saW5lL2Zvby5jc3MiLCJ0ZXN0L2ZpeHR1cmVzL2Rpc3QvZXh0cmFjdDo6c291cmNlbWFwOmlubGluZS9iYXIuY3NzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBO0VBQ0UsV0FBVztDQUNaOztBQ0ZEO0VBQ0UsV0FBVztDQUNaIiwiZmlsZSI6InRlc3QvZml4dHVyZXMvZGlzdC9leHRyYWN0Ojpzb3VyY2VtYXA6aW5saW5lL2J1bmRsZS5jc3MiLCJzb3VyY2VzQ29udGVudCI6WyJib2R5IHtcbiAgY29sb3I6IHJlZDtcbn1cbiIsIi5iYXIge1xuICBjb2xvcjogcmVkO1xufVxuIl19*/"
`;

exports[`extract::sourcemap:inline: js code 1`] = `
"'use strict';

console.log(undefined, undefined);
"
`;

exports[`extract::sourcemap:true: css code 1`] = `
"body {
  color: red;
}

.bar {
  color: red;
}

/*# sourceMappingURL=bundle.css.map */"
`;

exports[`extract::sourcemap:true: css map 1`] = `"{\\"version\\":3,\\"sources\\":[\\"test/fixtures/dist/extract::sourcemap:true/foo.css\\",\\"test/fixtures/dist/extract::sourcemap:true/bar.css\\"],\\"names\\":[],\\"mappings\\":\\"AAAA;EACE,WAAW;CACZ;;ACFD;EACE,WAAW;CACZ\\",\\"file\\":\\"test/fixtures/dist/extract::sourcemap:true/bundle.css\\",\\"sourcesContent\\":[\\"body {\\\\n  color: red;\\\\n}\\\\n\\",\\".bar {\\\\n  color: red;\\\\n}\\\\n\\"]}"`;

exports[`extract::sourcemap:true: js code 1`] = `
"'use strict';

console.log(undefined, undefined);
"
`;

exports[`inject:top: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"body {\\\\n  color: red;\\\\n}\\\\n\\";
__$$styleInject(css,{\\"insertAt\\":\\"top\\"});

var css$1 = \\".bar {\\\\n  color: red;\\\\n}\\\\n\\";
__$$styleInject(css$1,{\\"insertAt\\":\\"top\\"});

console.log(css, css$1);
"
`;

exports[`minimize:extract: css code 1`] = `
"body{color:red}
.bar{color:red}"
`;

exports[`minimize:extract: js code 1`] = `
"'use strict';

console.log(undefined, undefined);
"
`;

exports[`minimize:inject: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"body{color:red}\\";
__$$styleInject(css);

var css$1 = \\".bar{color:red}\\";
__$$styleInject(css$1);

console.log(css, css$1);
"
`;

exports[`modules:extract: css code 1`] = `
".style_foo__33weH {
  color: red;
}
"
`;

exports[`modules:extract: js code 1`] = `
"'use strict';

var style = {\\"foo\\":\\"style_foo__33weH\\"};

console.log(style.foo);
"
`;

exports[`modules:inject: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\".style_foo__33weH {\\\\n  color: red;\\\\n}\\\\n\\";
var style = {\\"foo\\":\\"style_foo__33weH\\"};
__$$styleInject(css);

console.log(style.foo);
"
`;

exports[`modules:named-exports: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var foo = \\"style_foo__23nAE\\";
var $new$ = \\"style_new__155Jg\\";
var css = \\".style_foo__23nAE {\\\\n  color: red;\\\\n}\\\\n\\\\n.style_new__155Jg {\\\\n  color: blue;\\\\n}\\\\n\\";
var style = {\\"foo\\":\\"style_foo__23nAE\\",\\"new\\":\\"style_new__155Jg\\"};
__$$styleInject(css);

var style$1 = Object.freeze({
	foo: foo,
	$new$: $new$,
	default: style
});

console.log(style$1);
"
`;

exports[`postcss-config: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"a {\\\\n  color: blue\\\\n}\\\\n\\\\n.multiline,\\\\n.selector {\\\\n  box-shadow: 1px 0 9px rgba(0, 0, 0, .4),\\\\n              1px 0 3px rgba(0, 0, 0, .6)\\\\n}\\\\n\\";
__$$styleInject(css);
"
`;

exports[`postcss-options: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"body {\\\\n  display: -webkit-box;\\\\n  display: -ms-flexbox;\\\\n  display: flex;\\\\n}\\\\n\\";
__$$styleInject(css);
"
`;

exports[`sass: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"#sidebar {\\\\n  width: 30%;\\\\n  background-color: #faa; }\\\\n\\";
__$$styleInject(css);
"
`;

exports[`sass:modules: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\".style_foo__2demv {\\\\n  color: red; }\\\\n\\";
var style = {\\"foo\\":\\"style_foo__2demv\\"};
__$$styleInject(css);

console.log(style);
"
`;

exports[`sass:sourcemap: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"#sidebar {\\\\n  width: 30%;\\\\n  background-color: #faa; }\\\\n\\\\n/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImZvby5zYXNzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBO0VBQ0UsV0FBVztFQUNYLHVCQUF1QixFQUFFIiwiZmlsZSI6ImZvby5zYXNzIiwic291cmNlc0NvbnRlbnQiOlsiI3NpZGViYXIge1xuICB3aWR0aDogMzAlO1xuICBiYWNrZ3JvdW5kLWNvbG9yOiAjZmFhOyB9XG4iXX0= */\\";
__$$styleInject(css);
"
`;

exports[`simple: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"body {\\\\n  color: red;\\\\n}\\\\n\\";
__$$styleInject(css);

var css$1 = \\".bar {\\\\n  color: red;\\\\n}\\\\n\\";
__$$styleInject(css$1);

console.log(css, css$1);
"
`;

exports[`skip-loader: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"body {\\\\n  color: red;\\\\n}\\\\n\\";
__$$styleInject(css);

var css$1 = \\".bar {\\\\n  color: red;\\\\n}\\\\n\\";
__$$styleInject(css$1);

console.log(css, css$1);
"
`;

exports[`sourcemap:inline: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"body {\\\\n  color: red;\\\\n}\\\\n\\\\n/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImZvby5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7RUFDRSxXQUFXO0NBQ1oiLCJmaWxlIjoiZm9vLmNzcyIsInNvdXJjZXNDb250ZW50IjpbImJvZHkge1xuICBjb2xvcjogcmVkO1xufVxuIl19 */\\";
__$$styleInject(css);

var css$1 = \\".bar {\\\\n  color: red;\\\\n}\\\\n\\\\n/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImJhci5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7RUFDRSxXQUFXO0NBQ1oiLCJmaWxlIjoiYmFyLmNzcyIsInNvdXJjZXNDb250ZW50IjpbIi5iYXIge1xuICBjb2xvcjogcmVkO1xufVxuIl19 */\\";
__$$styleInject(css$1);

console.log(css, css$1);
"
`;

exports[`sourcemap:true: js code 1`] = `
"'use strict';

function __$styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css) { return }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = \\"body {\\\\n  color: red;\\\\n}\\\\n\\\\n/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImZvby5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7RUFDRSxXQUFXO0NBQ1oiLCJmaWxlIjoiZm9vLmNzcyIsInNvdXJjZXNDb250ZW50IjpbImJvZHkge1xuICBjb2xvcjogcmVkO1xufVxuIl19 */\\";
__$$styleInject(css);

var css$1 = \\".bar {\\\\n  color: red;\\\\n}\\\\n\\\\n/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImJhci5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7RUFDRSxXQUFXO0NBQ1oiLCJmaWxlIjoiYmFyLmNzcyIsInNvdXJjZXNDb250ZW50IjpbIi5iYXIge1xuICBjb2xvcjogcmVkO1xufVxuIl19 */\\";
__$$styleInject(css$1);

console.log(css, css$1);
"
`;
